### ISSUE TITLE
Source file change detection and potential "back-dating" is sometimes relative to the past

### ISSUE THREAD

**Issue by schnittchen** [elixir-lang/elixir#8450]
### Environment

* Elixir & Erlang/OTP versions (elixir --version): 
```
Erlang/OTP 21 [erts-10.1.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe]

Elixir 1.7.4 (compiled with Erlang/OTP 21)
```
* Operating system: Elixir runs in docker container derived FROM elixir:1.7.4-slim, with Ubuntu 18.4 as host

### Current behavior

I am seeing this https://github.com/elixir-lang/elixir/blob/v1.6/lib/mix/lib/mix/utils.ex#L173 warning quite often under certain situations. Both source files and a few other files are reported.

For source files, the `File.touch!` happening subsequently means that my editor (vim) thinks the file has changed. After changing the file from the editor and recompiling, this can happen over again. This interacts with coding flow and makes for a bad experience.

So far I believe the behaviour

* is not related to Docker being involved (the container timezone corresponds to the host's)
  * I did not check what happens when running outside of Docker
* happens after resuming from suspend, so when OS time jumps forward
* goes away when restarting the VM (until it comes back after suspend)
  * I restarted the VM without restarting the Docker container

I suspect the time warping of system time lets system time lag behind OS time for a while. After the warning, the timestamp of the file as seen with `ls -l --time-style +%s` is BEFORE the current time as seen with `date +%s`.

<hr>

The code takes `System.system_time/1` as reference, and that behaviour was introduced here https://github.com/elixir-lang/elixir/commit/e0dc8b0051354c4d4da7cb391c683afbb5b343e6. I could not find more information on why that decision was taken, except that @michalmuskala suggested it, probably for a good reason.

**Comment by josevalim** [comment]
Thanks for the info.

Your VM is not supposed to be under time warped mode under regular
situations. If you start IEx manually and try the different
System.system_time/os_time functions, what are the differences between them
and the date command? How large is it?

The written file is meant to be behind date, but only for small amounts
(the amount it takes for you to run a command).

It is also important to rule out Docker. Can you reproduce it directly in
Ubuntu?
-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Director of R&D

**Comment by schnittchen** [comment]
After a longer suspend, this is on

```
Erlang/OTP 21 [erts-10.1.3] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe]

Elixir 1.7.4 (compiled with Erlang/OTP 20)
```

installed from the official packages, directly on the host:

```
iex(7)> {System.cmd("date", ["+%s"]),  System.system_time(:seconds)}
{{"1543787013\n", 0}, 1543780138}
```

**Comment by josevalim** [comment]
Something is definitely off here. It is reporting 6875 seconds difference!

Can you please call these:

```
{System.cmd("date", ["+%s"]),  System.system_time(:seconds), System.os_time(:seconds)}
:erlang.system_info(:os_monotonic_time_source)
:erlang.system_info(:os_system_time_source)
:erlang.system_info(:time_offset)
:erlang.system_info(:time_warp_mode)
:erlang.system_info(:time_correction)
:erlang.system_info(:start_time)
:erlang.system_info(:end_time)
```

?

**Comment by schnittchen** [comment]
I had shut down the VM, so I started a new one and suspended before this report (so time difference is far less than above):

```
iex(13)> {System.cmd("date", ["+%s"]),  System.system_time(:seconds), System.os_time(:seconds)}
{{"1543788481\n", 0}, 1543788230, 1543788481}
iex(14)> :erlang.system_info(:os_monotonic_time_source)
[
  function: :clock_gettime,
  clock_id: :CLOCK_MONOTONIC,
  resolution: 1000000000,
  extended: :no,
  parallel: :yes,
  time: 22128728091386
]
iex(15)> :erlang.system_info(:os_system_time_source)
[
  function: :clock_gettime,
  clock_id: :CLOCK_REALTIME,
  resolution: 1000000000, 
  parallel: :yes,
  time: 1543788481915074551
]
iex(16)> :erlang.system_info(:time_offset)
:final
iex(17)> :erlang.system_info(:time_warp_mode)
:no_time_warp
iex(18)> :erlang.system_info(:time_correction)
true
iex(19)> :erlang.system_info(:start_time)
-576460752000000000
iex(20)> :erlang.system_info(:end_time)
8646911283000000000
```

The value of `System.os_time(:seconds) - System.system_time(:seconds)` (~250) is very slowly decreasing, maybe by 1 or 2 per minute. (The value read is oscillating by a second, probably due to rounding))

**Comment by josevalim** [comment]
Right, at this point it is cleat that this is not an Elixir issue and we need to understand why the VM is immediately entering into time correction. It may not be an Erlang VM issue either.

What is your host OS? Searching for "clock drift docker" actually reveals possible issues, including this one: https://github.com/docker/for-mac/issues/2076

**Comment by schnittchen** [comment]
This is all on Ubuntu 18.04.

**Comment by josevalim** [comment]
Oh, that's much better if we can rule docker out.

**Comment by josevalim** [comment]
Do you see the same symptons as here (i.e. boot time changing?)? https://askubuntu.com/questions/1014285/excessive-system-clock-drift-2-minutes-per-hour

**Comment by schnittchen** [comment]
No, `cat /proc/stat | grep btime` is showing consistent output.

Also, `echo hpet > /sys/devices/system/clocksource/clocksource0/current_clocksource` did not change the fact that `System.os_time(:seconds) - System.system_time(:seconds)` returns a significant value after resuming from suspend.

For reference, I'm running `4.15.0-39-generic #42-Ubuntu SMP` Kernel, and (because of resuming being unreliable some time ago) I'm actually using hybrid-sleep (writes out to swap, then suspends) instead of simply suspending at the moment.

**Comment by schnittchen** [comment]
I've experienced it only after suspending so far.

**Comment by josevalim** [comment]
It would be nice to try to prove if the drift also happens at OS level. I assume this does not fix it? https://askubuntu.com/questions/881562/clock-problem-when-waking-from-suspend

But it seems to be an OS issue rather than Erlang one. Definitely not Elixir's.

**Comment by josevalim** [comment]
I will exceptionally ping @garazdawi because he may be able to confirm if this is a bug or not before we do report it to the Erlang/OTP team.

Lukas, to save you from reading the whole discussion, @schnittchen is experiencing clock drifting on Ubuntu 18.4 when he starts Elixir after his computer comes back from a suspension. In a nutshell, the Erlang VM would consider the current time (retrieved with `:erlang.universaltime`) to be somewhere in the past compared to the OS clock. You can see the results of many `system_info` when it happens in this [comment](https://github.com/elixir-lang/elixir/issues/8450#issuecomment-443547306). This makes us to touch files using a timestamp in the past (compared to the OS clock).

@garazdawi, can you let us know if there is a way for us to ask the OS its realtime and its monotonic time? Also, can you please confirm that `:erlang.universaltime` is supposed to return the os time? Thank you!

**Comment by garazdawi** [comment]
If the OS clock and `erlang:universaltime/0` differ then that is an OS bug/feature. `erlang:localtime/0` and `erlang:universaltime()` work with the glibc functions [gmtime](http://man7.org/linux/man-pages/man3/gmtime.3p.html) and [localtime](http://man7.org/linux/man-pages/man3/localtime.3p.html).

Are you sure that is it the function that is called?

**Comment by josevalim** [comment]
> Are you sure that is it the function that is called?

Yes:

https://github.com/elixir-lang/elixir/blob/ddc554a7efef67f16e462709e7f46662f21f7ab5/lib/mix/lib/mix/utils.ex#L168-L173

And `calendar:universal_time/0` calls `erlang:universaltime/0`. So what is happening is that when you save the file in your editor, you set a mtime, but because the VM is "running in the past", we end-up emitting a warning.

**Comment by garazdawi** [comment]
The only functions that we artificially drag should be `erlang:system_time/0`, `erlang:timestamp/0` and `erlang:now/0`. So I don't know why `erlang:universaltime/0` would return a time in the past.

**Comment by josevalim** [comment]
Thanks @garazdawi!

@schnittchen can you please let us know the result of this expression:

    {System.cmd("date", ["+%s"]),  System.system_time(:seconds), System.os_time(:seconds), :erlang.timestamp, :erlang.universaltime}

**Comment by schnittchen** [comment]
```
iex(1)> # before suspend/resume
nil
iex(2)> {System.cmd("date", ["+%s"]),  System.system_time(:seconds), System.os_time(:seconds), :erlang.timestamp, :erlang.universaltime}
{{"1544211222\n", 0}, 1544211222, 1544211222, {1544, 211222, 546321},
 {{2018, 12, 7}, {19, 33, 42}}}
iex(3)> # after suspend/resume
nil
iex(4)> {System.cmd("date", ["+%s"]),  System.system_time(:seconds), System.os_time(:seconds), :erlang.timestamp, :erlang.universaltime}
{{"1544211390\n", 0}, 1544211256, 1544211390, {1544, 211256, 586213},
 {{2018, 12, 7}, {19, 36, 30}}}
```

**Comment by josevalim** [comment]
@schnittchen Everything seems correct then. We are returning the same date as the one returned by `date` and that's what we use to write the files. Could it be that the one to blame is your editor that is rather setting a date in the future?

If you use `touch` on the file and compare it with the result of `date +%s`, is it in the future or not? Then, what if you do the same but saving through your editor instead of touching it?

**Comment by schnittchen** [comment]
> We are returning the same date as the one returned by date and that's what we use to write the files.

Not true after suspend/resume: the first and second tuple entry are synonymous in the `iex(2)>` line, but NOT in the `iex(4)>` line above.

And `System.system_time(:seconds)` is what is used as reference for source file checking.

**Comment by josevalim** [comment]
Oh, doh. When I was checking the code, I did through the commit you linked,
which points to the v1.6 branch. I am so sorry.

I think replacing system_time by os_time should be enough to fix this
issue. Thanks for pointing it out. You can try it out locally and let me
know. I will push a fix for this tomorrow.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by schnittchen** [comment]
Oh. I should have linked to the relevant version of the code... sorry.

Wow, I can just `Code.eval_file` the modified source and try!

Lest I forget and restart the VM, I will check the effect tomorrow.

**Comment by schnittchen** [comment]
`now = System.os_time(:second)` fixes it!

**Comment by josevalim** [comment]
Now in master, thank you.

**Comment by schnittchen** [comment]
It would be nice to have the one-line change backported to 1.7. I can provide a PR if that's ok.

**Comment by josevalim** [comment]
@schnittchen yes please!
